### PR TITLE
Add telemetry

### LINKED ISSUE #458: Add telemetry reporting
With https://github.com/redhat-developer/vscode-commons we can add telemetry report in to this extension.

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?
Add dependency to `redhat.vscode-commons` and send telemetry data.
It add hadler and pass all [telemetry_event](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#telemetry_event) from yaml-ls.

Also it send telemetry with any errors logged in LSP communication. 

### What issues does this PR fix or reference?
Resolve #458

### CODE REVIEW

**Review by JPinkney — COMMENTED** [review]
It looks like there's some inconsistent spacing between some methods, some do:

method1
method2

like here: https://github.com/redhat-developer/vscode-yaml/pull/490/files?file-filters%5B%5D=.ts#diff-87dc6d96fc959744de13764f38871c3357de6cb849052ca9a0806a8b4d7552f8R24

and other times its

method1

method2

like https://github.com/redhat-developer/vscode-yaml/pull/490/files?file-filters%5B%5D=.ts#diff-87dc6d96fc959744de13764f38871c3357de6cb849052ca9a0806a8b4d7552f8R60

File: package.json [hunk 1]
```diff
@@ -169,24 +172,31 @@
...
     "prettier": "^2.0.5",
     "rimraf": "^3.0.2",
+    "sinon": "^9.0.3",
+    "sinon-chai": "^3.5.0",
     "ts-node": "^3.3.0",
     "typescript": "4.1.2",
     "vscode-test": "^1.4.0"
   },
   "dependencies": {
+    "@redhat-developer/vscode-redhat-telemetry": "0.0.15",
```
Comment (fbricon) [inline package.json:199]: 0.0.18

File: src/extension.ts [hunk 2]
```diff
@@ -132,6 +150,8 @@ export function activate(context: ExtensionContext): SchemaExtensionAPI {
     client.onRequest(VSCodeContentRequest.type, (uri: string) => {
       return getJsonSchemaContent(uri, schemaCache);
     });
+
+    telemetry.send({ name: 'server_initialized', type: 'track' });
```
Comment (fbricon) [inline src/extension.ts:154]: yaml.server.initialized

Comment (fbricon) [reply src/extension.ts:154]: Actually `type` is optional, set to `track` by default in https://github.com/redhat-developer/vscode-redhat-telemetry/blob/main/src/services/reporter.ts#L35

File: src/telemetry.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+export class TelemetryErrorHandler implements ErrorHandler {
+  private restarts: number[] = [];
+  constructor(
+    private readonly telemetry: TelemetryService,
+    private readonly name: string,
+    private readonly maxRestartCount: number
+  ) {}
+
+  error(error: Error, message: Message, count: number): ErrorAction {
+    this.telemetry.send({ name: 'LSP Error', properties: { jsonrpc: message.jsonrpc, error: error.message } });
```
Comment (fbricon) [inline src/telemetry.ts:19]: name: 'yaml.lsp.error'

File: src/telemetry.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    this.delegate.append(value);
+  }
+  appendLine(value: string): void {
+    this.checkError(value);
+    this.delegate.appendLine(value);
+  }
+
+  private checkError(value: string): void {
+    if (value.startsWith('[Error') || value.startsWith('  Message: Request')) {
+      this.telemetry.send({ name: 'server_error', properties: { error_message: value } });
```
Comment (fbricon) [inline src/telemetry.ts:63]: name: 'yaml.server.error', properties: { error: value }

**Review by fbricon — CHANGES_REQUESTED** [review]

File: test/telemetry.test.ts [hunk 5]
```diff
@@ -80,12 +80,12 @@ describe('Telemetry Test', () => {
...
     it('should send telemetry if log error', () => {
       telemetryChannel.append('[Error Some');
-      expect(telemetry.send).calledOnceWith({ name: 'server_error', properties: { error_message: '[Error Some' } });
+      expect(telemetry.send).calledOnceWith({ name: 'yaml.server.error', properties: { error: '[Error Some' } });
     });
 
     it('should send telemetry if log error2', () => {
       telemetryChannel.appendLine('[Error Some');
-      expect(telemetry.send).calledOnceWith({ name: 'server_error', properties: { error_message: '[Error Some' } });
+      expect(telemetry.send).calledOnceWith({ name: 'yaml.server.error', properties: { error: '[Error Some' } });
```
Comment (fbricon) [inline test/telemetry.test.ts]: those are identical tests. what are you checking?

also my ocd kicks in when I see that missing closing bracket in `[Error Some`

File: test/telemetry.test.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    });
+
+    it('should send telemetry if log error', () => {
+      telemetryChannel.append('[Error Some');
+      expect(telemetry.send).calledOnceWith({ name: 'yaml.server.error', properties: { error: '[Error Some' } });
+    });
+
+    it('should send telemetry if log error2', () => {
+      telemetryChannel.appendLine('[Error Some');
+      expect(telemetry.send).calledOnceWith({ name: 'yaml.server.error', properties: { error: '[Error Some' } });
```
Comment (fbricon) [inline test/telemetry.test.ts:88]: My ocd kicks in when I see that missing closing bracket in `[Error Some` ;-)

**Review by fbricon — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -232,6 +232,10 @@ e.g.
 }
 ```
 
+## Data and Telemetry
+
+The `vscode-yaml` extension collects anonymous [usage data](USAGE_DATA.md) and sends it to Red Hat servers to help improve our products and services. Read our [privacy statement](https://developers.redhat.com/article/tool-data-collection) to learn more. This extension respects the `redhat.elemetry.enabled` setting which you can learn more about at https://github.com/redhat-developer/vscode-commons#how-to-disable-telemetry-reporting
```
Comment (joshuawilson) [inline README.md:237]: Is `redhat.elemetry.enabled` missing a `t`?

Comment (evidolob) [reply README.md:237]: That question for `vscode-commons`, as that property defined [there](https://github.com/redhat-developer/vscode-commons/blob/master/package.json#L29)

Comment (fbricon) [reply README.md:237]: @evidolob you made a typo `redhat.elemetry.enabled` should be `redhat.telemetry.enabled`

**Review by JPinkney — APPROVED** [review]

**Review by joshuawilson — APPROVED** [review]

**Pull request merged**